web/shared/src/typeahead.ts

```diff
@@ -48,14 +48,41 @@ export function remove_diacritics(s: string): string {
     return s.normalize("NFKD").replace(unicode_marks, "");
 }
 
+function upper_bound_lexicographically(prefix: string, words: string[]): number {
+    let left = 0;
+    let right = words.length;
+    let found = false;
+    while (left < right) {
+        const mid = Math.floor((left + right) / 2);
+        if (words[mid].startsWith(prefix)) {
+            left = mid + 1;
+            found = true;
+        } else if (words[mid] < prefix) {
+            left = mid + 1;
+        } else {
+            right = mid;
+        }
+    }
+    if (found) {
+        return left;
+    }
+    return 0;
+}
+
+function strcmp(a: string, b: string): number {
+    return a < b ? -1 : a > b ? 1 : 0;
+}
+
 // This function attempts to match a query with a source text.
 // * query is the user-entered search query
 // * source_str is the string we're matching in, e.g. a user's name
 // * split_char is the separator for this syntax (e.g. ' ').
+// * order_matters is an optional flag for if the order of words in the query matters.
```

Comment:
This seems like basically two different functions: one for the order_matters true case, and one where it's false.

The implementation and the call sites would probably both be clarified by making it two separate functions instead.

Reply:
Got it. Do you think there is need for a wrapper function for the resulting distinct functions? Their objective, after all, is the same.

Reply:
I think it'd be clearest for the call sites that want one thing to call the one function, and the call sites that want the other thing to call the other function.

Reply:
Have renamed the original function for clarity in a separate prep commit, and created a new function with the new code in the main commit.